File: java/day20_part1_2018.java

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class Main {
    static class Point {
        int x, y;
        Point(int x, int y) { this.x = x; this.y = y; }
        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Point)) return false;
            Point p = (Point) o;
            return x == p.x && y == p.y;
        }
        @Override
        public int hashCode() { return Objects.hash(x, y); }
    }

    static class DoorMap {
        Map<Point, Map<Point, Boolean>> map = new HashMap<>();
        void addDoor(Point from, Point to) {
            map.computeIfAbsent(from, k -> new HashMap<>()).put(to, true);
        }
        Map<Point, Boolean> getDoors(Point p) { return map.getOrDefault(p, Collections.emptyMap()); }
    }

    public static void main(String[] args) throws IOException {
        String regex = new String(Files.readAllBytes(Paths.get("input.txt"))).trim();
        DoorMap dm = buildMap(regex.substring(1, regex.length() - 1));
        System.out.println(findFurthestRoom(dm));
    }

    static DoorMap buildMap(String regex) {
        DoorMap dm = new DoorMap();
        Stack<Point> stack = new Stack<>();
        Point cp = new Point(0, 0);
        for (char c : regex.toCharArray()) {
            switch (c) {
                case '(' -> stack.push(cp);
                case '|' -> cp = stack.peek();
                case ')' -> cp = stack.pop();
                default -> {
                    Point np = move(cp, c);
                    dm.addDoor(cp, np);
                    cp = np;
                }
            }
        }
        return dm;
    }

    static Point move(Point p, char dir) {
        return switch (dir) {
            case 'N' -> new Point(p.x, p.y - 1);
            case 'S' -> new Point(p.x, p.y + 1);
            case 'E' -> new Point(p.x + 1, p.y);
            case 'W' -> new Point(p.x - 1, p.y);
            default -> p;
        };
    }

    static int findFurthestRoom(DoorMap dm) {
        Map<Point, Integer> visited = new HashMap<>();
        Queue<Point> queue = new LinkedList<>();
        queue.add(new Point(0, 0));
        visited.put(new Point(0, 0), 0);
        int maxDoors = 0;

        while (!queue.isEmpty()) {
            Point p = queue.poll();
            for (Point np : dm.getDoors(p).keySet()) {
                if (!visited.containsKey(np)) {
                    visited.put(np, visited.get(p) + 1);
                    maxDoors = Math.max(maxDoors, visited.get(np));
                    queue.add(np);
                }
            }
        }
        return maxDoors;
    }
}

File: typescript/day20_part1_2018.ts
import * as fs from 'fs';

interface Point {
  x: number;
  y: number;
}

interface DoorMap {
  [key: string]: { [key: string]: boolean };
}

const regex = fs.readFileSync('input.txt', 'utf8');
const dm = buildMap(regex.slice(1, -1));
const maxDoors = findFurthestRoom(dm);
console.log(maxDoors);

function buildMap(regex: string): DoorMap {
  const dm: DoorMap = {};
  const stack: Point[] = [];
  let cp: Point = { x: 0, y: 0 };
  for (const c of regex) {
    if (c === '(') {
      stack.push(cp);
    } else if (c === '|') {
      cp = stack[stack.length - 1];
    } else if (c === ')') {
      cp = stack[stack.length - 1];
      stack.pop();
    } else {
      const np = move(cp, c);
      if (!dm[`${cp.x},${cp.y}`]) {
        dm[`${cp.x},${cp.y}`] = {};
      }
      dm[`${cp.x},${cp.y}`][`${np.x},${np.y}`] = true;
      cp = np;
    }
  }
  return dm;
}

function move(p: Point, dir: string): Point {
  switch (dir) {
    case 'N':
      return { x: p.x, y: p.y - 1 };
    case 'S':
      return { x: p.x, y: p.y + 1 };
    case 'E':
      return { x: p.x + 1, y: p.y };
    case 'W':
      return { x: p.x - 1, y: p.y };
    default:
      return p;
  }
}

function findFurthestRoom(dm: DoorMap): number {
  const visited: { [key: string]: number } = {};
  const queue: Point[] = [{ x: 0, y: 0 }];
  let maxDoors = 0;

  while (queue.length > 0) {
    const p = queue.shift() as Point;
    for (const np in dm[`${p.x},${p.y}`]) {
      if (!visited[np]) {
        visited[np] = (visited[`${p.x},${p.y}`] || 0) + 1;
        maxDoors = Math.max(maxDoors, visited[np]);
        queue.push({ x: parseInt(np.split(',')[0]), y: parseInt(np.split(',')[1]) });
      }
    }
  }
  return maxDoors;
}